core/src/main/java/org/allenai/scienceparse/Parser.java

```diff
@@ -928,13 +928,28 @@ public ExtractedMetadata doParseWithTimeout(final InputStream is, final long tim
     final int parseNumber = nextParseNumber.getAndIncrement();
 
     final Thread t = Thread.currentThread();
-    final TimerTask task = new TimerTask() {
+
+    final TimerTask killTaskHard = new TimerTask() {
       @Override
       public void run() {
         synchronized (parseNumbersInProgress) {
-          if(parseNumbersInProgress.remove(parseNumber)) {
+          if(parseNumbersInProgress.contains(parseNumber)) {
+            logger.info("Killing parsing thread {} because it's taking too long", t.getId());
+            t.stop();
+            // I know this is dangerous. This is a last resort.
+          }
+        }
+      }
+    };
+
+    final TimerTask killTaskSoftly = new TimerTask() {
+      @Override
+      public void run() {
+        synchronized (parseNumbersInProgress) {
+          if(parseNumbersInProgress.contains(parseNumber)) {
             logger.info("Interrupting parsing thread {} because it's taking too long", t.getId());
             t.interrupt();
+            parserKillerTimer.schedule(killTaskHard, 2*timeoutInMs);

```

Comment:
I think a flat structure where you schedule both killers outside the task would be easier to read.

Reply:
I made it so.